# Data Wrangling

<h5>

**Wrangling** /ˈræŋ.ɡəl.ɪŋ/

the activity of taking care of, controlling, or moving animals, especially large animals such as cows or horses

</h5>

([Cambridge Dictionary](https://dictionary.cambridge.org/dictionary/english/wrangling))

![Cattle Wrangler - image from https://commons.wikimedia.org/wiki/File:Pioneer_Day_Wrangler.jpg](https://upload.wikimedia.org/wikipedia/commons/thumb/8/83/Pioneer_Day_Wrangler.jpg/320px-Pioneer_Day_Wrangler.jpg)

**[Data wrangling](https://en.wikipedia.org/wiki/Data_wrangling)** commonly refers to the transformation of data from one "input" format (e.g., `.csv` files from an experiment), to a different format (e.g., a tidy dataframe) that is more appropriate to the needs of an analysis. In the context of the ExPra experiments, you will use data wrangling techniques to implement the transformations and data cleaning steps specified in your preregistrations.

## Setup

### Setup Part 1: Install Packages

We will use [`tidyverse`](https://www.tidyverse.org/) packages to implement our data wrangling. The "tidyverse" is a series of packages which share a philosphy based around code and data structures that are (a) tidy, and (b) readable. These packages are frequently based around tidy dataframes, known as "[tibbles](https://tibble.tidyverse.org/)". You can install the tidyverse packages like so:

```
install.packages("tidyverse")
```

This includes many packages that we won't be using today, but which will be useful in other parts of the course (e.g., on Data Visualisation).

Remember, you should install packages in the console - never in a script that your share with others. This is because otherwise, your script will go to the effort of reinstalling a package *every time* it is run!

### Setup Part 2: Check the Packages Load

Now, we can test that the packages we will be using today actually load. You should be able to run this code without any errors:

In [1]:
options(repr.plot.width=3.5, repr.plot.height=3, repr.matrix.max.rows=10)

In [2]:
library(dplyr)
library(tidyr)


Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




The [`dplyr`](https://dplyr.tidyverse.org/) package has useful functions for manipulating tibbles (dataframes), such as sorting, filtering, and editing columns. The [`tidyr`](https://tidyr.tidyverse.org/) package has functions that help us tidy or reformat dataframes.

<img src="https://www.tidyverse.org/css/images/hex/dplyr.png" width=138>
<img src="https://www.tidyverse.org/css/images/hex/tidyr.png" width=138>

### Setup Part 3: Check the Data Loads

Finally, check that you can access the dataset we'll be using in this session. The `starwars` dataset is a dataset built into R that contains details of characters from the Star Wars films:

In [3]:
print(starwars)

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk~    172    77 blond      fair       blue            19   male  mascu~
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu~
 3 R2-D2        96    32 NA         white, bl~ red             33   none  mascu~
 4 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
 5 Leia Or~    150    49 brown      light      brown           19   fema~ femin~
 6 Owen La~    178   120 brown, gr~ light      blue            52   male  mascu~
 7 Beru Wh~    165    75 brown      light      blue            47   fema~ femin~
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu~
 9 Biggs D~    183    84 black      light      brown           24   male  mascu~
10 Obi-Wan~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~
# i 77 m

This snapshot shows an example of tidy data - a philosophy of organising data such that each observation (*character*) has a single row, with all variables tied to that character as a single column.

Now that we're all set up, let's start a-wrangling...

## `arrange()`

We can use the `arrange()` function to sort by variables in the dataframe. For example, we can arrange all characters in order of height (shortest to tallest) like so:

In [4]:
arrange(starwars, height)

name                  height mass hair_color skin_color  eye_color
1   Yoda                  66     17   white      green       brown    
2   Ratts Tyerell         79     15   none       grey, blue  unknown  
3   Wicket Systri Warrick 88     20   brown      brown       brown    
4   Dud Bolt              94     45   none       blue, grey  yellow   
5   R2-D2                 96     32   NA         white, blue red      
... ...                   ...    ...  ...        ...         ...      
83  Finn                  NA     NA   black      dark        dark     
84  Rey                   NA     NA   brown      light       hazel    
85  Poe Dameron           NA     NA   brown      light       brown    
86  BB8                   NA     NA   none       none        black    
87  Captain Phasma        NA     NA   unknown    unknown     unknown  
    birth_year sex    gender    homeworld   species       
1   896        male   masculine NA          Yoda's species
2    NA        male   masculine Aleen Minor Aleena        
3     8        male   masculine Endor       Ewok          
4    NA        male   masculine Vulpter     Vulptereen    
5    33        none   masculine Naboo       Droid         
... ...        ...    ...       ...         ...           
83  NA         male   masculine NA          Human         
84  NA         female feminine  NA          Human         
85  NA         male   masculine NA          Human         
86  NA         none   masculine NA          Droid         
87  NA         NA     NA        NA          NA            
    films                                                                                                                                                                        
1   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi                                                       
2   The Phantom Menace                                                                                                                                                           
3   Return of the Jedi                                                                                                                                                           
4   The Phantom Menace                                                                                                                                                           
5   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
... ...                                                                                                                                                                          
83  The Force Awakens                                                                                                                                                            
84  The Force Awakens                                                                                                                                                            
85  The Force Awakens                                                                                                                                                            
86  The Force Awakens                                                                                                                                                            
87  The Force Awakens                                                                                                                                                            
    vehicles starships          
1                               
2                               
3                               
4                               
5                               
... ...      ...                
83                              
84                              
85         

In [5]:
arrange(starwars, height) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yoda         66    17 white      green      brown            896 male  mascu~
 2 Ratts T~     79    15 none       grey, blue unknown           NA male  mascu~
 3 Wicket ~     88    20 brown      brown      brown              8 male  mascu~
 4 Dud Bolt     94    45 none       blue, grey yellow            NA male  mascu~
 5 R2-D2        96    32 NA         white, bl~ red               33 none  mascu~
 6 R4-P17       96    NA none       silver, r~ red, blue         NA none  femin~
 7 R5-D4        97    32 NA         white, red red               NA none  mascu~
 8 Sebulba     112    40 none       grey, red  orange            NA male  mascu~
 9 Gasgano     122    NA none       white, bl~ black             NA male  mascu~
10 Watto       137    NA black      blue, grey yellow            NA male  mascu~
# i 77 m

The `arrange()` function, like most `dplyr` verb functions, takes the dataframe (`starwars`) as its first argument, and the variable names (e.g., `height`) as subsequent arguments.

In height order, we can see that Yoda is the shortest character, at 66 cm, with podracer [Ratts Tyerell](https://starwars.fandom.com/wiki/Ratts_Tyerell) next shortest, at 79 cm.

We can also arrange the dataframe in *descending* order, with the *`desc()`* function.

In [6]:
arrange(starwars, desc(height))

name           height mass hair_color skin_color eye_color birth_year
1   Yarael Poof    264     NA  none       white      yellow     NA       
2   Tarfful        234    136  brown      brown      blue       NA       
3   Lama Su        229     88  none       grey       black      NA       
4   Chewbacca      228    112  brown      unknown    blue      200       
5   Roos Tarpals   224     82  none       grey       orange     NA       
... ...            ...    ...  ...        ...        ...       ...       
83  Finn           NA     NA   black      dark       dark      NA        
84  Rey            NA     NA   brown      light      hazel     NA        
85  Poe Dameron    NA     NA   brown      light      brown     NA        
86  BB8            NA     NA   none       none       black     NA        
87  Captain Phasma NA     NA   unknown    unknown    unknown   NA        
    sex    gender    homeworld species 
1   male   masculine Quermia   Quermian
2   male   masculine Kashyyyk  Wookiee 
3   male   masculine Kamino    Kaminoan
4   male   masculine Kashyyyk  Wookiee 
5   male   masculine Naboo     Gungan  
... ...    ...       ...       ...     
83  male   masculine NA        Human   
84  female feminine  NA        Human   
85  male   masculine NA        Human   
86  none   masculine NA        Droid   
87  NA     NA        NA        NA      
    films                                                                                                                      
1   The Phantom Menace                                                                                                         
2   Revenge of the Sith                                                                                                        
3   Attack of the Clones                                                                                                       
4   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
5   The Phantom Menace                                                                                                         
... ...                                                                                                                        
83  The Force Awakens                                                                                                          
84  The Force Awakens                                                                                                          
85  The Force Awakens                                                                                                          
86  The Force Awakens                                                                                                          
87  The Force Awakens                                                                                                          
    vehicles starships                           
1                                                
2                                                
3                                                
4   AT-ST    Millennium Falcon, Imperial shuttle 
5                                                
... ...      ...                                 
83                                               
84                                               
85           T-70 X-wing fighter                 
86                                               
87

In [7]:
arrange(starwars, desc(height)) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yarael ~    264    NA none       white      yellow          NA   male  mascu~
 2 Tarfful     234   136 brown      brown      blue            NA   male  mascu~
 3 Lama Su     229    88 none       grey       black           NA   male  mascu~
 4 Chewbac~    228   112 brown      unknown    blue           200   male  mascu~
 5 Roos Ta~    224    82 none       grey       orange          NA   male  mascu~
 6 Grievous    216   159 none       brown, wh~ green, y~       NA   male  mascu~
 7 Taun We     213    NA none       grey       black           NA   fema~ femin~
 8 Rugor N~    206    NA none       green      orange          NA   male  mascu~
 9 Tion Me~    206    80 none       grey       black           NA   male  mascu~
10 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
# i 77 m

This shows us that long-necked Jedi, [Yarael Poof](https://starwars.fandom.com/wiki/Yarael_Poof), is the tallest character, at 264 cm.

As well as sorting by numeric variables, we can also sort by character variables. For instance, we can sort by hair colour. If we do this we will sort hair alphabetically, from "auburn" to "white".

In [8]:
arrange(starwars, hair_color)

name                  height mass hair_color    skin_color       eye_color
1   Mon Mothma            150      NA auburn        fair             blue     
2   Wilhuff Tarkin        180      NA auburn, grey  fair             blue     
3   Obi-Wan Kenobi        182    77.0 auburn, white fair             blue-gray
4   Biggs Darklighter     183    84.0 black         light            brown    
5   Boba Fett             183    78.2 black         fair             brown    
... ...                   ...    ...  ...           ...              ...      
83  C-3PO                 167      75 NA            gold             yellow   
84  R2-D2                  96      32 NA            white, blue      red      
85  R5-D4                  97      32 NA            white, red       red      
86  Greedo                173      74 NA            green            black    
87  Jabba Desilijic Tiure 175    1358 NA            green-tan, brown orange   
    birth_year sex            gender    homeworld species
1   48.0       female         feminine  Chandrila Human  
2   64.0       male           masculine Eriadu    Human  
3   57.0       male           masculine Stewjon   Human  
4   24.0       male           masculine Tatooine  Human  
5   31.5       male           masculine Kamino    Human  
... ...        ...            ...       ...       ...    
83  112        none           masculine Tatooine  Droid  
84   33        none           masculine Naboo     Droid  
85   NA        none           masculine Tatooine  Droid  
86   44        male           masculine Rodia     Rodian 
87  600        hermaphroditic masculine Nal Hutta Hutt   
    films                                                                                                                                                                        
1   Return of the Jedi                                                                                                                                                           
2   Revenge of the Sith, A New Hope                                                                                                                                              
3   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
4   A New Hope                                                                                                                                                                   
5   The Empire Strikes Back, Attack of the Clones   , Return of the Jedi                                                                                                         
... ...                                                                                                                                                                          
83  The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
84  The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
85  A New Hope                                                                                                                                                                   
86  A New Hope                                                                                                                                                                   
87  The Phantom Menace, Return of the Jedi, A New Hope                                                                                                                           
    vehicles       
1                  
2                  
3   Tribubble bongo
4                  
5                  
... ...            
83                 
84                 
85                 
86                 
87  

In [9]:
arrange(starwars, hair_color) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Mon Mot~    150  NA   auburn     fair       blue            48   fema~ femin~
 2 Wilhuff~    180  NA   auburn, g~ fair       blue            64   male  mascu~
 3 Obi-Wan~    182  77   auburn, w~ fair       blue-gray       57   male  mascu~
 4 Biggs D~    183  84   black      light      brown           24   male  mascu~
 5 Boba Fe~    183  78.2 black      fair       brown           31.5 male  mascu~
 6 Lando C~    177  79   black      dark       brown           31   male  mascu~
 7 Watto       137  NA   black      blue, grey yellow          NA   male  mascu~
 8 Quarsh ~    183  NA   black      dark       brown           62   NA    NA    
 9 Shmi Sk~    163  NA   black      fair       brown           72   fema~ femin~
10 Eeth Ko~    171  NA   black      brown      brown           NA   male  mascu~
# i 77 m

Finally, we can sort by multiple variables at once with comma-separated statements. For instance, we may want to sort by hair colour alphabetically, and then by height descendingly:

In [10]:
arrange(starwars, hair_color, desc(height))

name                  height mass hair_color    skin_color       eye_color
1   Mon Mothma            150    NA   auburn        fair             blue     
2   Wilhuff Tarkin        180    NA   auburn, grey  fair             blue     
3   Obi-Wan Kenobi        182    77   auburn, white fair             blue-gray
4   Bail Prestor Organa   191    NA   black         tan              brown    
5   Gregar Typho          185    85   black         dark             brown    
... ...                   ...    ...  ...           ...              ...      
83  Jabba Desilijic Tiure 175    1358 NA            green-tan, brown orange   
84  Greedo                173      74 NA            green            black    
85  C-3PO                 167      75 NA            gold             yellow   
86  R5-D4                  97      32 NA            white, red       red      
87  R2-D2                  96      32 NA            white, blue      red      
    birth_year sex            gender    homeworld species
1   48         female         feminine  Chandrila Human  
2   64         male           masculine Eriadu    Human  
3   57         male           masculine Stewjon   Human  
4   67         male           masculine Alderaan  Human  
5   NA         male           masculine Naboo     Human  
... ...        ...            ...       ...       ...    
83  600        hermaphroditic masculine Nal Hutta Hutt   
84   44        male           masculine Rodia     Rodian 
85  112        none           masculine Tatooine  Droid  
86   NA        none           masculine Tatooine  Droid  
87   33        none           masculine Naboo     Droid  
    films                                                                                                                                                                        
1   Return of the Jedi                                                                                                                                                           
2   Revenge of the Sith, A New Hope                                                                                                                                              
3   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
4   Attack of the Clones, Revenge of the Sith                                                                                                                                    
5   Attack of the Clones                                                                                                                                                         
... ...                                                                                                                                                                          
83  The Phantom Menace, Return of the Jedi, A New Hope                                                                                                                           
84  A New Hope                                                                                                                                                                   
85  The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
86  A New Hope                                                                                                                                                                   
87  The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
    vehicles       
1                  
2                  
3   Tribubble bongo
4                  
5                  
... ...            
83                 
84                 
85                 
86                 
87  

In [11]:
arrange(starwars, hair_color, desc(height)) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Mon Mot~    150  NA   auburn     fair       blue            48   fema~ femin~
 2 Wilhuff~    180  NA   auburn, g~ fair       blue            64   male  mascu~
 3 Obi-Wan~    182  77   auburn, w~ fair       blue-gray       57   male  mascu~
 4 Bail Pr~    191  NA   black      tan        brown           67   male  mascu~
 5 Gregar ~    185  85   black      dark       brown           NA   male  mascu~
 6 Biggs D~    183  84   black      light      brown           24   male  mascu~
 7 Boba Fe~    183  78.2 black      fair       brown           31.5 male  mascu~
 8 Quarsh ~    183  NA   black      dark       brown           62   NA    NA    
 9 Jango F~    183  79   black      tan        brown           66   male  mascu~
10 Lando C~    177  79   black      dark       brown           31   male  mascu~
# i 77 m

### Check your Knowledge!

Try coming up with code to solve the following:

1. Sort descendingly by mass, such that the most massive character comes first.

2. Sort by gender, eye colour, and then height. Which character is the first observation? Why?

## `filter()`

The `filter()` function is useful for subsetting data. For example, we can easily filter our dataset to find all our data on Darth Vader. The following code says that we should *filter* the dataframe called *starwars* to only include rows where the variable called `name` has the value `"Darth Vader"`:

In [12]:
filter(starwars, name=="Darth Vader")

name        height mass hair_color skin_color eye_color birth_year sex 
1 Darth Vader 202    136  none       white      yellow    41.9       male
  gender    homeworld species
1 masculine Tatooine  Human  
  films                                                                                             
1 The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             
  vehicles starships      
1          TIE Advanced x1

In [13]:
filter(starwars, name=="Darth Vader") |> print()

# A tibble: 1 x 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Darth Va~    202   136 none       white      yellow          41.9 male  mascu~
# i 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


<img src="https://media1.giphy.com/media/Uu4WP50jNo1uZeor4t/giphy.gif" width=250>

<sub><sup>[via giphy](https://media.giphy.com/media/Uu4WP50jNo1uZeor4t/giphy.gif)</sup></sub>

To find all characters that are *not* Darth Vader, we can use `!=`, which stands for "does not equal." This returns all rows in the dataframe where the `name` column does not have the value `"Darth Vader"`.

In [14]:
filter(starwars, name!="Darth Vader")

name           height mass hair_color  skin_color  eye_color birth_year
1   Luke Skywalker 172     77  blond       fair        blue       19       
2   C-3PO          167     75  NA          gold        yellow    112       
3   R2-D2           96     32  NA          white, blue red        33       
4   Leia Organa    150     49  brown       light       brown      19       
5   Owen Lars      178    120  brown, grey light       blue       52       
... ...            ...    ...  ...         ...         ...       ...       
82  Rey             NA    NA   brown       light       hazel     NA        
83  Poe Dameron     NA    NA   brown       light       brown     NA        
84  BB8             NA    NA   none        none        black     NA        
85  Captain Phasma  NA    NA   unknown     unknown     unknown   NA        
86  Padmé Amidala  165    45   brown       light       brown     46        
    sex    gender    homeworld species
1   male   masculine Tatooine  Human  
2   none   masculine Tatooine  Droid  
3   none   masculine Naboo     Droid  
4   female feminine  Alderaan  Human  
5   male   masculine Tatooine  Human  
... ...    ...       ...       ...    
82  female feminine  NA        Human  
83  male   masculine NA        Human  
84  none   masculine NA        Droid  
85  NA     NA        NA        NA     
86  female feminine  Naboo     Human  
    films                                                                                                                                                                        
1   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens                                                        
2   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
3   The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
4   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens                                                        
5   Attack of the Clones, Revenge of the Sith , A New Hope                                                                                                                       
... ...                                                                                                                                                                          
82  The Force Awakens                                                                                                                                                            
83  The Force Awakens                                                                                                                                                            
84  The Force Awakens                                                                                                                                                            
85  The Force Awakens                                                                                                                                                            
86  Attack of the Clones, The Phantom Menace  , Revenge of the Sith                                                                                                              
    vehicles                                    
1   Snowspeeder          , Imperial Speeder Bike
2                                               
3                                               
4   Imperial Speeder Bike                       
5                                               
... ...                                         
82                                              
83                                              
84                                             

In [15]:
filter(starwars, name!="Darth Vader") |> print()

# A tibble: 86 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk~    172    77 blond      fair       blue            19   male  mascu~
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu~
 3 R2-D2        96    32 NA         white, bl~ red             33   none  mascu~
 4 Leia Or~    150    49 brown      light      brown           19   fema~ femin~
 5 Owen La~    178   120 brown, gr~ light      blue            52   male  mascu~
 6 Beru Wh~    165    75 brown      light      blue            47   fema~ femin~
 7 R5-D4        97    32 NA         white, red red             NA   none  mascu~
 8 Biggs D~    183    84 black      light      brown           24   male  mascu~
 9 Obi-Wan~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~
10 Anakin ~    188    84 blond      fair       blue            41.9 male  mascu~
# i 76 m

We can also filter to include a list of characters. To do this, we first define a character vector of characters we wish to keep. We can then filter to only include characters whose `name` is in (`%in%`) that vector:

In [16]:
cool_droids <- c("C-3PO", "R2-D2", "IG-88")
filter(starwars, name %in% cool_droids)

name  height mass hair_color skin_color  eye_color birth_year sex  gender   
1 C-3PO 167     75  NA         gold        yellow    112        none masculine
2 R2-D2  96     32  NA         white, blue red        33        none masculine
3 IG-88 200    140  none       metal       red        15        none masculine
  homeworld species
1 Tatooine  Droid  
2 Naboo     Droid  
3 NA        Droid  
  films                                                                                                                                                                        
1 The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
2 The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
3 The Empire Strikes Back                                                                                                                                                      
  vehicles starships
1                   
2                   
3

In [17]:
filter(starwars, name %in% cool_droids) |> print()

# A tibble: 3 x 14
  name  height  mass hair_color skin_color  eye_color birth_year sex   gender   
  <chr>  <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>    
1 C-3PO    167    75 NA         gold        yellow           112 none  masculine
2 R2-D2     96    32 NA         white, blue red               33 none  masculine
3 IG-88    200   140 none       metal       red               15 none  masculine
# i 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


As with `==` and `!=`, we can invert `%in%` to only include characters who are *not* in the list. To do this, we put an exclamation mark (`!`) at the *start* of the statement:

In [18]:
filter(starwars, !(name %in% cool_droids))

name               height mass hair_color  skin_color eye_color birth_year
1   Luke Skywalker     172     77  blond       fair       blue      19.0      
2   Darth Vader        202    136  none        white      yellow    41.9      
3   Leia Organa        150     49  brown       light      brown     19.0      
4   Owen Lars          178    120  brown, grey light      blue      52.0      
5   Beru Whitesun lars 165     75  brown       light      blue      47.0      
... ...                ...    ...  ...         ...        ...       ...       
80  Rey                 NA    NA   brown       light      hazel     NA        
81  Poe Dameron         NA    NA   brown       light      brown     NA        
82  BB8                 NA    NA   none        none       black     NA        
83  Captain Phasma      NA    NA   unknown     unknown    unknown   NA        
84  Padmé Amidala      165    45   brown       light      brown     46        
    sex    gender    homeworld species
1   male   masculine Tatooine  Human  
2   male   masculine Tatooine  Human  
3   female feminine  Alderaan  Human  
4   male   masculine Tatooine  Human  
5   female feminine  Tatooine  Human  
... ...    ...       ...       ...    
80  female feminine  NA        Human  
81  male   masculine NA        Human  
82  none   masculine NA        Droid  
83  NA     NA        NA        NA     
84  female feminine  Naboo     Human  
    films                                                                                                                      
1   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
2   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
3   The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
4   Attack of the Clones, Revenge of the Sith , A New Hope                                                                     
5   Attack of the Clones, Revenge of the Sith , A New Hope                                                                     
... ...                                                                                                                        
80  The Force Awakens                                                                                                          
81  The Force Awakens                                                                                                          
82  The Force Awakens                                                                                                          
83  The Force Awakens                                                                                                          
84  Attack of the Clones, The Phantom Menace  , Revenge of the Sith                                                            
    vehicles                                    
1   Snowspeeder          , Imperial Speeder Bike
2                                               
3   Imperial Speeder Bike                       
4                                               
5                                               
... ...                                         
80                                              
81                                              
82                                              
83                                              
84                                              
    starships                                                    
1   X-wing          , Imperial shuttle                           
2   TIE Advanced x1                                              
3                                                                
4                                                                
5                                                                
... ...                                                          
80

In [19]:
filter(starwars, !(name %in% cool_droids)) |> print()

# A tibble: 84 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk~    172    77 blond      fair       blue            19   male  mascu~
 2 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
 3 Leia Or~    150    49 brown      light      brown           19   fema~ femin~
 4 Owen La~    178   120 brown, gr~ light      blue            52   male  mascu~
 5 Beru Wh~    165    75 brown      light      blue            47   fema~ femin~
 6 R5-D4        97    32 NA         white, red red             NA   none  mascu~
 7 Biggs D~    183    84 black      light      brown           24   male  mascu~
 8 Obi-Wan~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~
 9 Anakin ~    188    84 blond      fair       blue            41.9 male  mascu~
10 Wilhuff~    180    NA auburn, g~ fair       blue            64   male  mascu~
# i 74 m

The `filter()` function can also deal with numeric values. For instance, we can filter to only include characters who are shorter than, or are exactly, 96 cm tall. To do this we use `<=`, which stands for "less than or equal to", or "≤".

In [20]:
filter(starwars, height<=96)

name                  height mass hair_color skin_color  eye_color birth_year
1 R2-D2                 96     32   NA         white, blue red        33       
2 Yoda                  66     17   white      green       brown     896       
3 Wicket Systri Warrick 88     20   brown      brown       brown       8       
4 Dud Bolt              94     45   none       blue, grey  yellow     NA       
5 Ratts Tyerell         79     15   none       grey, blue  unknown    NA       
6 R4-P17                96     NA   none       silver, red red, blue  NA       
  sex  gender    homeworld   species       
1 none masculine Naboo       Droid         
2 male masculine NA          Yoda's species
3 male masculine Endor       Ewok          
4 male masculine Vulpter     Vulptereen    
5 male masculine Aleen Minor Aleena        
6 none feminine  NA          Droid         
  films                                                                                                                                                                        
1 The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
2 The Empire Strikes Back, Attack of the Clones   , The Phantom Menace     , Revenge of the Sith    , Return of the Jedi                                                       
3 Return of the Jedi                                                                                                                                                           
4 The Phantom Menace                                                                                                                                                           
5 The Phantom Menace                                                                                                                                                           
6 Attack of the Clones, Revenge of the Sith                                                                                                                                    
  vehicles starships
1                   
2                   
3                   
4                   
5                   
6

In [21]:
filter(starwars, height<=96) |> print()

# A tibble: 6 x 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 R2-D2         96    32 NA         white, bl~ red               33 none  mascu~
2 Yoda          66    17 white      green      brown            896 male  mascu~
3 Wicket S~     88    20 brown      brown      brown              8 male  mascu~
4 Dud Bolt      94    45 none       blue, grey yellow            NA male  mascu~
5 Ratts Ty~     79    15 none       grey, blue unknown           NA male  mascu~
6 R4-P17        96    NA none       silver, r~ red, blue         NA none  femin~
# i 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


Finally, as in other `dplyr` functions, we can combine multiple comma-separated statements in one use of the `filter()` function. Here we filter to only include characters who:
* come from Tatooine
* are at least 100 cm tall
* are human

In [22]:
filter(starwars, homeworld=="Tatooine", height>=100, species=="Human")

name               height mass hair_color  skin_color eye_color birth_year
1 Luke Skywalker     172     77  blond       fair       blue      19.0      
2 Darth Vader        202    136  none        white      yellow    41.9      
3 Owen Lars          178    120  brown, grey light      blue      52.0      
4 Beru Whitesun lars 165     75  brown       light      blue      47.0      
5 Biggs Darklighter  183     84  black       light      brown     24.0      
6 Anakin Skywalker   188     84  blond       fair       blue      41.9      
7 Shmi Skywalker     163     NA  black       fair       brown     72.0      
8 Cliegg Lars        183     NA  brown       fair       blue      82.0      
  sex    gender    homeworld species
1 male   masculine Tatooine  Human  
2 male   masculine Tatooine  Human  
3 male   masculine Tatooine  Human  
4 female feminine  Tatooine  Human  
5 male   masculine Tatooine  Human  
6 male   masculine Tatooine  Human  
7 female feminine  Tatooine  Human  
8 male   masculine Tatooine  Human  
  films                                                                                                                      
1 The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope             , The Force Awakens      
2 The Empire Strikes Back, Revenge of the Sith    , Return of the Jedi     , A New Hope                                      
3 Attack of the Clones, Revenge of the Sith , A New Hope                                                                     
4 Attack of the Clones, Revenge of the Sith , A New Hope                                                                     
5 A New Hope                                                                                                                 
6 Attack of the Clones, The Phantom Menace  , Revenge of the Sith                                                            
7 Attack of the Clones, The Phantom Menace                                                                                   
8 Attack of the Clones                                                                                                       
  vehicles                                    
1 Snowspeeder          , Imperial Speeder Bike
2                                             
3                                             
4                                             
5                                             
6 Zephyr-G swoop bike, XJ-6 airspeeder        
7                                             
8                                             
  starships                                                                   
1 X-wing          , Imperial shuttle                                          
2 TIE Advanced x1                                                             
3                                                                             
4                                                                             
5 X-wing                                                                      
6 Trade Federation cruiser, Jedi Interceptor        , Naboo fighter           
7                                                                             
8

In [23]:
filter(starwars, homeworld=="Tatooine", height>=100, species=="Human") |> print()

# A tibble: 8 x 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Luke Sky~    172    77 blond      fair       blue            19   male  mascu~
2 Darth Va~    202   136 none       white      yellow          41.9 male  mascu~
3 Owen Lars    178   120 brown, gr~ light      blue            52   male  mascu~
4 Beru Whi~    165    75 brown      light      blue            47   fema~ femin~
5 Biggs Da~    183    84 black      light      brown           24   male  mascu~
6 Anakin S~    188    84 blond      fair       blue            41.9 male  mascu~
7 Shmi Sky~    163    NA black      fair       brown           72   fema~ femin~
8 Cliegg L~    183    NA brown      fair       blue            82   male  mascu~
# i 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>
